server.go

```diff
@@ -3726,6 +3727,51 @@ func (s *server) connectToPeer(addr *lnwire.NetAddress,
 	s.OutboundPeerConnected(nil, conn)
 }
 
+// updateConnRequests checks to see if there are any existing connection requests
+// active for the given pubKey and if there are then it checks that there are
+// connection requests active for all the given addresses. If any are missing
+// then new connection requests are created.
+func (s *server) updateConnRequests(pubKey *btcec.PublicKey, addrs []net.Addr) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Check if there are currently any connection requests for this node.
```

Comment:
Something I don't fully understand is that isn't this inherently a race? As in if the address changes we only ever connect to that new address if we are right in the middle of connecting to the node anyway?

Reply:
Ah ok yeah is your question basically 'what happens if we somehow get the node announcement before we run `peerTerminationWatcher`? 

For some reason im having hard time recreating this situation in the itest (cant seem to force the NodeAnnouncement to be received before peerTerminationWatcher has run no matter where i add delays 🤔 )

To address this though, what do you think about:
1) removing the `if p.Inbound()` [here](https://github.com/lightningnetwork/lnd/blob/master/server.go#L3500) so that latest addresses are fetched from node announcement no matter if peer was inbound or outbound
2) also then fetching latest advertised addresses again after the backoff [here](https://github.com/lightningnetwork/lnd/blob/master/server.go#L3564)

we would still need to keep the UpdateConnReqs though for the case where NodeAnnouncement is received after peerTerminationWatcher has run

Reply:
> removing the if p.Inbound() here so that latest addresses are fetched from node announcement no matter if peer was inbound or outbound
also then fetching latest advertised addresses again after the backoff here

Sounds good to me.

> we would still need to keep the UpdateConnReqs though for the case where NodeAnnouncement is received after peerTerminationWatcher has run

Not sure that I follow. What is the problem if `NodeAnnouncement` is received after `peerTerminationWatcher`?

Reply:
peerTerminationWatcher creates the connection requests that will be used to reconnect to the peer and uses the currently stored addresses for that node. If the peer changes their address, but we only get their NodeAnnouncement that includes the new addresses after we have run peerTerminationWatcher, then the connection requests that we created will have the incorrect addresses

Reply:
I see. Then I think we need to fetch the node info from db again to get the latest address.

Reply:
well yes, but those latest addresses could still be old. We could get a NodeAnnouncement after we have done that. That is why I added the: `r.cfg.UpdateConnRequests(pk, msg.Addresses)` above so that if we do get a NodeAnnouncement after we have already fetched from the db and created connReqs, then things can be updated. I will also add refresh from db in peerTerminationWatcher, but i dont think I can remove the `UpdateConnRequests`. Thoughts?

Reply:
> I will also add refresh from db in peerTerminationWatcher

Yeah I think this would be good enough. As suggested in [this comment](https://github.com/lightningnetwork/lnd/issues/5377#issuecomment-861857325), we should handle persistent connections on their own dedicated scope. Down the bottom,`connmgr` handles the retries, but it does not refresh the address. Since `peerTerminationWatcher` is a wrapper above `connmgr`, maybe we could put some logic there to retrieve the updated address?

Unrelated to this PR, I think we need to someday extract all the connection management into its own package, saving us from the long `server.go` file, put up some nice unit tests, etc.

Reply:
I have updated the PR and the PR description to explain why i think it is not enough to just refresh from the db in peerTerminationWatcher (the itest also fails if we only do this). Let me know what you think.